java/source/flight.rst

```diff
@@ -0,0 +1,485 @@
+    import org.apache.arrow.vector.types.pojo.Field;
+    import org.apache.arrow.vector.types.pojo.FieldType;
+    import org.apache.arrow.vector.types.pojo.Schema;
+
+    import java.io.IOException;
+    import java.nio.charset.StandardCharsets;
+    import java.util.ArrayList;
+    import java.util.Arrays;
+    import java.util.Collections;
+    import java.util.Iterator;
+    import java.util.List;
+    import java.util.concurrent.ConcurrentHashMap;
+
+    class DataInMemory {
+        private List<ArrowRecordBatch> listArrowRecordBatch;
+        private Schema schema;
+        private long rows;
+        public DataInMemory(List<ArrowRecordBatch> listArrowRecordBatch, Schema schema, long rows) {
+            this.listArrowRecordBatch = listArrowRecordBatch;
+            this.schema = schema;
+            this.rows = rows;
+        }
+        public List<ArrowRecordBatch> getListArrowRecordBatch() {
+            return listArrowRecordBatch;
+        }
+        public Schema getSchema() {
+            return schema;
+        }
+        public long getRows() {
+            return rows;
+        }
+    }
+
+    // Server
+    Location location = Location.forGrpcInsecure("0.0.0.0", 33333);
+    try (RootAllocator allocator = new RootAllocator(Long.MAX_VALUE)){
+        class CookbookProducer extends NoOpFlightProducer {
+            ConcurrentHashMap<FlightDescriptor, DataInMemory> dataInMemory = new ConcurrentHashMap<>();
+            @Override
+            public Runnable acceptPut(CallContext context, FlightStream flightStream, StreamListener<PutResult> ackStream) {
+                List<ArrowRecordBatch> listArrowRecordBatch = new ArrayList<>();
+                return () -> {
+                    long rows = 0;
+                    while (flightStream.next()) {
+                        VectorUnloader unloader = new VectorUnloader(flightStream.getRoot());
+                        try (final ArrowRecordBatch arb = unloader.getRecordBatch()) {
+                            listArrowRecordBatch.add(arb);
+                            rows += flightStream.getRoot().getRowCount();
+                        }
+                    }
+                    DataInMemory pojoFlightDataInMemory = new DataInMemory(listArrowRecordBatch, flightStream.getSchema(), rows);
+                    dataInMemory.put(flightStream.getDescriptor(), pojoFlightDataInMemory);
+                    ackStream.onCompleted();
+                };
+            }
+
+            @Override
+            public void getStream(CallContext context, Ticket ticket, ServerStreamListener listener) {
+                FlightDescriptor flightDescriptor = FlightDescriptor.path(new String(ticket.getBytes(), StandardCharsets.UTF_8)); // Recover data for key configured
+                if(dataInMemory.containsKey(flightDescriptor)){
+                    VectorSchemaRoot vectorSchemaRoot = VectorSchemaRoot.create(dataInMemory.get(flightDescriptor).getSchema(), allocator);
```

Comment:
basically just do 

```
var dataset = dataInMemory.get(flightDescriptor);
if (dataset == null) { // raise an error }
// use dataset...
```

otherwise you have race conditions

Reply:
Added